### PR TITLE
Fix grafana links for rancher workload deployment 

### PR DESCRIPTION
**Pull request by skanakal**
### Summary
Fixes https://github.com/rancher/dashboard/issues/10236

### Occurred changes and/or fixed issues

Added a condition to the `replicaSetId` getter method in the Deployment class to accurately fetch the latest replicaSetId

`relationship.message.split(':')[1].trim() > 0;`

### Technical notes summary

The issue seems to be that the get replicaSetId [method](https://github.com/rancher/dashboard/blob/master/shell/models/apps.deployment.js#L14) is not accurately identifying the correct ReplicaSet associated with the latest Deployment update. Let's say if you have multiple ReplicaSets, the logic needs to ensure it fetches the most recent one. Typically, k8s assigns ReplicaSets with a revision number, where a higher number indicates a newer ReplicaSet but we're not using them here, the current logic only fetches the first replica set id from the array... not the one with replicaset 1/1 running.
 
grafana_url from dashboard fetching the wrong replicasetid...
```
https://master-0/api/v1/namespaces/cattle-monitoring-system/services/http:rancher-monitoring-grafana:80/proxy/d/rancher-workload-pods-1/rancher-workload-pods?orgId=1&from=now-5m&to=now&refresh=30s&var-namespace=default&var-kind=ReplicaSet&var-workload=httpd-5bc8dd79b&theme=dark   

root@master-0:~# k get rs  | grep httpd
httpd-5bc8dd79b    0         0         0       5h25m
httpd-745cfdfb84   0         0         0       5h25m
httpd-7d99dfc997   1         1         1       106m
root@master-0:~# 
root@master-0:~# 
root@master-0:~# k get deployment httpd -o jsonpath='{.status.conditions[0].message} {"\n"}'
ReplicaSet "httpd-7d99dfc997" has successfully progressed. 
root@master-0:~#
```

image

### Areas or cases that should be tested
- Install the latest version of Rancher.
- Deploy the latest version of the Monitoring App.
- Create nginx deployement using the `nginx:1.12.0` image.
- Create httpd application using the `httpd:latest` image.
- Upgrade the httpd deployment to the `httpd:alpine` image.
- Validate that the metrics displayed for httpd correspond to the httpd deployment.

### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone 
- [x] The PR template has been filled out
- [x] The PR has been self reviewed 
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes

### CODE REVIEW

File: shell/models/apps.deployment.js [hunk 1]
```diff
@@ -14,7 +14,8 @@ export default class Deployment extends Workload {
   get replicaSetId() {
     const set = this.metadata?.relationships?.find((relationship) => {
       return relationship.rel === 'owner' &&
-            relationship.toType === WORKLOAD_TYPES.REPLICA_SET;
+            relationship.toType === WORKLOAD_TYPES.REPLICA_SET &&
+            relationship.message.split(':')[1].trim() > 0;
```
Comment (richard-cox) [inline shell/models/apps.deployment.js:18]: This seems fragile, if the message does not contain `:` it'll throw an exception. It feels like we need to check all relationships and if one does contain the required text return the condition, if none contain the required text return the first one as before.

**Review by richard-cox — CHANGES_REQUESTED** [review]
One comment, please can you also run through the checklist

File: shell/models/apps.deployment.js (hunk 1 above)
Comment (skanakal) [reply shell/models/apps.deployment.js:18]: I've updated the code to handle this case safely... Here is the modified one. 

```
export default class Deployment extends Workload {
  get replicaSetId() {
    const relationships = this.metadata?.relationships || [];
    
    // Find all relevant ReplicaSet relationships
    const replicaSetRelationships = relationships.filter((relationship) => 
      relationship.rel === 'owner' && relationship.toType === WORKLOAD_TYPES.REPLICA_SET
    );

    // Filter the ReplicaSets based on replicas > 0
    const activeReplicaSet = replicaSetRelationships.find((relationship) => {
      const replicasMatch = relationship.message?.match(/Replicas: (\d+)/);
      const replicas = replicasMatch ? parseInt(replicasMatch[1], 10) : 0;
      return replicas > 0;
    });

    // If no active ReplicaSet is found, fall back to the first one from the list
    const selectedReplicaSet = activeReplicaSet || replicaSetRelationships[0];

    return selectedReplicaSet?.toId?.replace(`${this.namespace}/`, '');
  }
```

Comment (richard-cox) [reply shell/models/apps.deployment.js:18]: TBH this looks good. There's probably a way to do it in a single loop with `reduce` but there's not going to be a lot of looping or frequent calls.

I would suggest though moving the regex out to a `const` at the top of the file. Those can sometime initialise slowly so it's good practise reduce if possible.

const replicasRegEx = /.../
...
.match(replicasRegEx)

**Comment by richard-cox** [comment]
As per discussion

1) (extra) might be worth a rebase against rancher/dashboard `master`. this might resolve the failing storybook ci gate
2) update given comments above
3) i did have a quick go at some unit tests, given this it might be best to include before merging. would you be able to finish off? 

shell/models/__tests__/apps.deployment.test.ts
```
import Deployment from '@shell/models/apps.deployment';
import { WORKLOAD_TYPES } from '@shell/config/types';

describe('class Deployment', () => {
  describe('replicateSetId', () => {
    it.each([{
      relationships: [],
      expected:      undefined,
    }, {
      relationships: [{
        rel:    'owner',
        toType: WORKLOAD_TYPES.REPLICA_SET,
        toId:   'rel-id'
      }],
      expected: `rel-id`,
    }])('replicateSetID', ({ relationships, expected }) => {
      const deploymentData = {
        id:       'any-id',
        type:     WORKLOAD_TYPES.DEPLOYMENT,
        metadata: {
          name:      'any-name',
          namespace: 'any-namespace',
          uid:       'any-uid',
          relationships,
        },
      };

      const deployment = new Deployment(deploymentData);

      expect(deployment.replicaSetId).toStrictEqual(expected);
    });
  });
});

```

Notes
- tests can be run with `yarn test`
- Additional test cases can be added to the array in `it.each`
- cases should cover 
  - single relationship entry with a valid Replicas: x value (higher than zero)
  - multiple relationships with last one with a valid Replicas: x value (higher than zero)
  - multiple relationships with all invalid Replicas: 0
  - multiple relationships with messages that don't contain Replicas: x
  - any other angles you can think of

File: shell/models/apps.deployment.js (hunk 1 above)
Comment (skanakal) [reply shell/models/apps.deployment.js]: Thank you for the suggestion! As recommended, I’ve moved the `replicasRegEx` constant to the top of the file.

**Review by richard-cox — APPROVED** [review]

**Pull request merged**